system_test/enterprise_e2e/tests/test_net_client.py

```
@@ -199,3 +202,377 @@ def test_no_request_durable_ack_loses_rows(server_factory, net_sidecar,
+    regression (e.g. the demoted node NACKing instead of sending a reconnect-
+    eligible role-change close), so surface it as a descriptive assertion."""
+    try:
+        net_sidecar.send(_CLUSTER_TABLE, count=count, start_index=start_index)
+        return net_sidecar.flush()
+    except RuntimeError as e:  # NetSidecarError is a RuntimeError
+        raise AssertionError(
+            f"store-and-forward producer hard-failed while ingesting rows "
+            f"[{start_index}..{start_index + count}) across a role transition — a "
+            f"graceful role change must surface to the sender as a reconnect-eligible "
+            f"close (retry from SF), never a terminal; only SF exhaustion may be "
+            f"terminal. sidecar error: {e!r}")
```

Ruff (0.15.21)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)